### PR TITLE
feat: get compiled job assets from JobRunInput api

### PR DESCRIPTION
**Pull request by Mryashbhardwaj**
proton PR:: https://github.com/odpf/proton/pull/155

Future plans : https://github.com/odpf/optimus/issues/443

### CODE REVIEW

File: api/handler/v1beta1/job_run.go [hunk 1]
```diff
@@ -140,14 +141,73 @@ func (sv *JobRunServiceServer) RegisterInstance(ctx context.Context, req *pb.Reg
...
-		Context: &pb.InstanceContext{
+		Context: &pb.JobRunInputResponse{
 			Envs:    jobRunInput.ConfigMap,
 			Secrets: jobRunInput.SecretsMap,
 			Files:   jobRunInput.FileMap,
 		},
 	}, nil
 }
 
+func (sv *JobRunServiceServer) GetJobRunInput(ctx context.Context, req *pb.RegisterInstanceRequest) (*pb.JobRunInputResponse, error) {
```
Comment (sbchaos) [inline api/handler/v1beta1/job_run.go:152]: Have we added a test for this ?

Comment (Mryashbhardwaj) [reply api/handler/v1beta1/job_run.go]: have added now

File: api/handler/v1beta1/job_run.go [hunk 2]
```diff
@@ -140,14 +141,74 @@ func (sv *JobRunServiceServer) RegisterInstance(ctx context.Context, req *pb.Reg
...
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "%s: failed to get JobRun by jobRunId::%s for job %s ", err.Error(), req.JobrunId, jobSpec.Name)
+		}
+	}
+	secrets, err := sv.secretService.GetSecrets(ctx, namespaceSpec)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "%s: failed to get secrets %s:%s", err.Error(), jobSpec.Name, namespaceSpec.Name)
+	}
+
+	jobRunInput, err := sv.jobRunInputCompiler.CompileNewJobSpec(ctx,
```
Comment (sravankorumilli) [inline api/handler/v1beta1/job_run.go:194]: want to have a small discussion to understand what are the inputs for compilation

File: service/monitoring_service.go [hunk 3]
```diff
@@ -17,14 +22,33 @@ type monitoringService struct {
 
 type MonitoringService interface {
 	ProcessEvent(context.Context, models.JobEvent, models.NamespaceSpec, models.JobSpec) error
+	GetJobRunByScheduledAt(context.Context, models.NamespaceSpec, models.JobSpec, time.Time) (models.JobRunSpec, error)
+	GetJobRunByRunID(context.Context, uuid.UUID) (models.JobRunSpec, error)
+}
+
+func (m monitoringService) GetJobRunByScheduledAt(ctx context.Context, namespaceSpec models.NamespaceSpec, jobSpec models.JobSpec, scheduledAt time.Time) (models.JobRunSpec, error) {
+	return m.JobRunMetricsRepository.GetActiveJobRun(ctx, scheduledAt.Format(store.ISODateFormat), namespaceSpec, jobSpec)
```
Comment (sravankorumilli) [inline service/monitoring_service.go:30]: it is not just active job run it should fetch even for old job runs so the method name contradicts.

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

**Comment by sravankorumilli** [comment]
I think it is better to create another PR for deprecating the old admin build instance command and the corresponding api

File: cmd/job/run_input.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	assetOutputDir string
+	runType        string
+	runName        string
+	scheduledAt    string
+
+	keysWithUnsubstitutedValue []string
+}
+
+// NewJobRunInputCommand gets compiled assets required for a job run
+func NewJobRunInputCommand(clientConfig *config.ClientConfig) *cobra.Command {
```
Comment (sravankorumilli) [inline cmd/job/run_input.go:43]: what is the response when job run doesn't exist

File: store/store.go [hunk 5]
```diff
@@ -100,10 +104,11 @@ type JobRunRepository interface {
 
 // JobRunMetricsRepository represents a storage interface for Job runs generated to
 type JobRunMetricsRepository interface {
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply store/store.go:106]: the reason is that, we haven't still deprecated the old "job_run" repository hence needed a new name for the new table repository

File: store/store.go [hunk 6]
```diff
@@ -100,10 +104,11 @@ type JobRunRepository interface {
...
-	Save(context.Context, models.JobEvent, models.NamespaceSpec, models.JobSpec, int64) error
+	Save(context.Context, models.JobEvent, models.NamespaceSpec, models.JobSpec, int64, string) error
 	Update(context.Context, models.JobEvent, models.NamespaceSpec, models.JobSpec) error
 	Get(context.Context, models.JobEvent, models.NamespaceSpec, models.JobSpec) (models.JobRunSpec, error)
 	GetActiveJobRun(context.Context, string, models.NamespaceSpec, models.JobSpec) (models.JobRunSpec, error)
+	GetByID(context.Context, uuid.UUID) (models.JobRunSpec, error)
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply store/store.go:111]: All `Get` methods return `JobRunSpec` 
This one does not need `JobEvent` as it fetched by ID

File: store/postgres/job_run_repository.go [hunk 7]
```diff
@@ -50,12 +53,22 @@ func (JobRunMetrics) TableName() string {
 	return "job_run"
 }
 
+func jsonToInstanceSpec(j datatypes.JSON) []models.InstanceSpecData {
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply store/postgres/job_run_repository.go:56]: sure

File: api/handler/v1beta1/job_run.go (hunk 2 above)
Comment (Mryashbhardwaj) [reply api/handler/v1beta1/job_run.go:194]: sure

File: cmd/job/run_input.go (hunk 4 above)
Comment (Mryashbhardwaj) [reply cmd/job/run_input.go:43]: `Error: request failed for job <job_name>: rpc error: code = Internal desc = resource not found: failed to get JobRun by ScheduledAt for job <job_name>`

File: cmd/job/run_input.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	cmd.MarkFlagRequired("output-dir")
+	cmd.Flags().StringVar(&jobRunInput.scheduledAt, "scheduled-at", "", "Time at which the job was scheduled for execution")
+	cmd.MarkFlagRequired("scheduled-at")
+	cmd.Flags().StringVar(&jobRunInput.runType, "type", "task", "Type of instance, could be task/hook")
+	cmd.MarkFlagRequired("type")
+	cmd.Flags().StringVar(&jobRunInput.runName, "name", "", "Name of instancetype")
+	cmd.MarkFlagRequired("name")
+
+	cmd.Flags().StringP("project-name", "p", defaultProjectName, "Name of the optimus project")
+	cmd.MarkFlagRequired("project-name")
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply cmd/job/run_input.go:70]: done

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

File: store/postgres/job_run_repository.go [hunk 9]
```diff
@@ -136,17 +176,49 @@ func (repo *JobRunMetricsRepository) Get(ctx context.Context, event models.JobEv
...
+func (repo *JobRunMetricsRepository) Save(ctx context.Context, event models.JobEvent, namespaceSpec models.NamespaceSpec, jobSpec models.JobSpec, slaMissDurationInSec int64, jobDestination string) error {
 	eventPayload := event.Value
 
-	scheduledAtTimeStamp, err := time.Parse(ISODateFormat, eventPayload["scheduled_at"].GetStringValue())
+	scheduledAtTimeStamp, err := time.Parse(store.ISODateFormat, eventPayload["scheduled_at"].GetStringValue())
+
+	if err != nil {
+		return err
+	}
+	executedAt := time.Unix(int64(eventPayload["task_start_timestamp"].GetNumberValue()), 0)
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply store/postgres/job_run_repository.go:195]: `task_start_timestamp` can be changed to `job_run_start_timestamp`

[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply store/postgres/job_run_repository.go:195]: > with this executedAt changes for every task_run, lets check whether this is expected behaviour or not.

no, this changes for each job_run_attempt (i.e. job_start_event)

File: store/postgres/job_run_repository.go [hunk 10]
```diff
@@ -136,17 +176,49 @@ func (repo *JobRunMetricsRepository) Get(ctx context.Context, event models.JobEv
...
+
+	instanceSpec := models.InstanceSpec{
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply store/postgres/job_run_repository.go:197]: will avoid persisting this

File: service/monitoring_service.go [hunk 11]
```diff
@@ -17,31 +22,50 @@ type monitoringService struct {
...
+
+func (m monitoringService) GetJobRunByScheduledAt(ctx context.Context, namespaceSpec models.NamespaceSpec, jobSpec models.JobSpec, scheduledAt time.Time) (models.JobRunSpec, error) {
+	return m.JobRunMetricsRepository.GetLatestJobRunByScheduledTime(ctx, scheduledAt.Format(store.ISODateFormat), namespaceSpec, jobSpec)
+}
+func (m monitoringService) GetJobRunByRunID(ctx context.Context, jobRunID uuid.UUID) (models.JobRunSpec, error) {
+	return m.JobRunMetricsRepository.GetByID(ctx, jobRunID)
 }
 
 func (m monitoringService) registerNewJobRun(ctx context.Context, event models.JobEvent, namespaceSpec models.NamespaceSpec, jobSpec models.JobSpec) error {
+	var jobDestination string
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply service/monitoring_service.go:37]: sure

File: ext/scheduler/airflow2/resources/__lib.py [hunk 12]
```diff
@@ -395,8 +395,8 @@ def optimus_notify(context, event_meta):
         
         "exception": str(context.get('exception')) or "",
 
-        "scheduled_at"          : current_execution_date.strftime(TIMESTAMP_FORMAT),
-        "scheduled_at_ts"       : datetime.timestamp(context.get('execution_date')),
+        "scheduled_at"          : schedule_time.strftime(TIMESTAMP_FORMAT),
+        "scheduled_at_ts"       : datetime.timestamp(schedule_time),
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply ext/scheduler/airflow2/resources/__lib.py:399]: just 2 diff formats, can be handled on the server as well, can remove this

**Review by Mryashbhardwaj — COMMENTED** [review]
ready for review

File: Makefile [hunk 13]
```diff
@@ -5,7 +5,7 @@ NAME = "github.com/odpf/optimus"
 LAST_COMMIT := $(shell git rev-parse --short HEAD)
 LAST_TAG := "$(shell git rev-list --tags --max-count=1)"
 OPMS_VERSION := "$(shell git describe --tags ${LAST_TAG})-next"
-PROTON_COMMIT := "23765066adf5704d2df026cbc65a98e45f415cf7"
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply Makefile]: done

**Review by sravankorumilli — APPROVED** [review]

**Pull request merged**